src/middlewared/middlewared/plugins/domain_controller.py

```diff
@@ -0,0 +1,79 @@
+from freenasUI.common.samba import Samba4
+from freenasUI.common.system import get_hostname
+
+from middlewared.schema import accepts, Dict, Int, Str
+from middlewared.service import private, SystemServiceService
+
+
+class DomainControllerService(SystemServiceService):
+
+    class Config:
+        service = "domaincontroller"
+        datastore_extend = "domaincontroller.domaincontroller_extend"
+        datastore_prefix = "dc_"
+
+    @private
+    async def domaincontroller_extend(self, domaincontroller):
+        domaincontroller['role'] = domaincontroller['role'].upper()
+        return domaincontroller
+
+    @private
+    async def domaincontroller_compress(self, domaincontroller):
+        domaincontroller['role'] = domaincontroller['role'].lower()
+        return domaincontroller
+
+    @accepts(Dict(
+        'domaincontroller_update',
+        Str('realm'),
+        Str('domain'),
+        Str('role', enum=["DC"]),
+        Str('dns_backend', enum=["SAMBA_INTERNAL", "BIND9_FLATFILE", "BIND9_DLZ", "NONE"]),
+        Str('dns_forwarder'),
+        Str('forest_level', enum=["2000", "2003", "2008", "2008_R2", "2012", "2012_R2"]),
+        Str('passwd'),
+        Int('kerberos_realm', required=False),
+    ))
+    async def do_update(self, data):
+        old = await self.config()
+
+        new = old.copy()
+        new.update(data)
+
+        if new["kerberos_realm"] is None:
+            hostname = get_hostname()
+            dc_hostname = "%s.%s" % (hostname, new["realm"].lower())
```

Comment:
Would you mind pet my OCD and transform it to f-string or format()? I want to get rid of old formatting in middleware ; )